src/hotspot/share/prims/jvmtiImpl.cpp

```diff
@@ -962,10 +962,10 @@ JvmtiDeferredEvent JvmtiDeferredEventQueue::dequeue() {
 }
 
 void JvmtiDeferredEventQueue::post(JvmtiEnv* env) {
-  // Post and destroy queue nodes
+  // Post events while nmethods are still in the queue and can't be unloaded or made zombie
   while (_queue_head != NULL) {
-     JvmtiDeferredEvent event = dequeue();
-     event.post_compiled_method_load_event(env);
+    _queue_head->event().post_compiled_method_load_event(env);
+    dequeue();
```

Comment:
Good find!  So we _can_ zombie the nmethod after we take it off the list.  Makes a lot of sense.  Thank you for your perseverance in tracking this down!